doc/developer/design/20240411_graceful_reconfiguration_of_managed_clusters_v1.md

```diff
@@ -0,0 +1,302 @@
+this, and we add some big disclaimers to clusters with sources.
+
+### Hydration-Based Scheduling
+There is currently the beginning of the mechanism to look at hydration
+based metrics for scheduling in `compute-client/src/controller/
+instance:update_hydration_status`, this works very well for a single data flow,
+but can't gauge the hydration status of a replica as a whole.
+
+What we likely want is a mechanism [discussed here](https://github.com/MaterializeInc/materialize/issues/20010#issuecomment-2058563052), where we
+compare the data-flow frontiers between reconfiguration and active replicas,
+and when the reconfiguration replicas have all data-flows within some threshold
+of the active, we send a claim that the check condition has been met and proceed with
+finalization of the reconfiguration.
+
+__Potential challenges with using hydration status__:
+It is difficult to know when "hydration" has finished because:
+  - New data coming in may change the timestamp and lead to hydration status inaccuracies
+  - New data-flows can be created ad-hock which will need to be "hydrated"
+  - hydration_status errors related to Environmentd restart?
+    (https://materializeinc.slack.com/archives/CTESPM7FU/p1712227236046369?thread_ts=1712224682.263469&cid=CTESPM7FU)
+Some of these can be resolved by only looking at data flows that were installed
+when the replica was created. It seems like comparing active and reconfiguration
+replicas will likely work better than attempting to gauge when hydration has
+finished for a given replica. It also may allow us to shorten the timeout if all
+replicas for a cluster undergoing reconfiguration are restarted.
+
+For hydration-based scheduling the `until caught up` syntax will be used.
+
+
+### Reconfiguration replica naming:
+Is it still useful to have reconfiguration replica naming differences if we also have a table tracking those
+replicas? It seems like we could get the list just as easily with a join and skip a step in the reconfiguration task.
+
+### Should the reconfiguration task create the replicas
+Currently, I have the reconfiguration task set to be in charge of the wait check and the finalization, but not
+the reconfiguration replica creation nor the insertion into `mz_cluster_pending_replicas`. This seems like a reasonable
+separation, as if we cannot do these two things we should fail and not background a task.
+
+
+### Do we want a more generic jobs table?
+The proposed mechanism is for a sort of one-off very specific job engine. Assuming we
+want a more robust job engine in the future it may make sense to take baby steps towards that here.
+For instance, we could introduce an `mz_job` table that stores all information about a ReconfiguratTask. This
+could be used on environmentd restart to re-initialize the task or perform cleanup for the task if the envd restart
+would've caused the task to fail. I haven't thought this through much.
```

Comment:
I'm extremely 👍🏽 on doing something one off for altering clusters (i.e., the design you've proposed), vs building a more generic jobs engine. CockroachDB has a generic jobs engine, and it's not terribly clean; different types of jobs have significantly different characteristics.